src/bin/pg_autoctl/primary_standby.c

```diff
@@ -1036,6 +1043,209 @@ primary_rewind_to_standby(LocalPostgresServer *postgres)
 }
 
 
+/*
+ * postgres_maybedo_crash_recovery implements a round of Postgres crash
+ * recovery for the local instance of Postgres when pg_rewind would otherwise
+ * fail because of its internal checks.
+ */
+bool
+postgres_maybe_do_crash_recovery(LocalPostgresServer *postgres)
+{
+	PostgresSetup *pgSetup = &(postgres->postgresSetup);
+	ReplicationSource *replicationSource = &(postgres->replicationSource);
+
+	LocalExpectedPostgresStatus *pgStatus = &(postgres->expectedPgStatus);
+
+	/* update our service controller for Postgres to release control */
+	if (!keeper_set_postgres_state_unknown(&(pgStatus->state),
+										   pgStatus->pgStatusPath))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/* we don't log the output for pg_ctl_status here */
+	int status = pg_ctl_status(pgSetup->pg_ctl, pgSetup->pgdata, false);
+
+	if (status != PG_CTL_STATUS_NOT_RUNNING)
+	{
+		log_error("Failed to prepare for crash recovery: "
+				  "Postgres is not stopped");
+		return false;
+	}
+
+	/*
+	 * pg_rewind fails when the target cluster (meaning the local Postgres
+	 * instance) is either running or has not been shutdown correctly. Time to
+	 * use pg_controldata and see if the DBState there is to pg_rewind liking.
+	 */
+	const bool missingPgdataIsOk = false;
+
+	if (!pg_controldata(pgSetup, missingPgdataIsOk))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/*
+	 * We know that Postgres is not running thanks to pg_ctl_status, and we
+	 * just grabbed the output from pg_controldata. We can now implement the
+	 * same pre-condition checks as in Postgres pg_rewind.c.
+	 */
+	if (pgSetup->control.state != DB_SHUTDOWNED &&
+		pgSetup->control.state != DB_SHUTDOWNED_IN_RECOVERY)
```

Comment:
If we invert this check, then we can use an early `return true`.

Reply:
It is meant to mimic Postgres pg_rewind test that reads:

``` C
	if (!no_ensure_shutdown &&
		ControlFile_target.state != DB_SHUTDOWNED &&
		ControlFile_target.state != DB_SHUTDOWNED_IN_RECOVERY)
```

Note that the `--no-ensure-shutdown` is new in Postgres 13, and I just noticed that now.